File: src/main/java/ltdjms/discord/shop/services/ShopAdminNotificationService.java
package ltdjms.discord.shop.services;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.product.domain.Product;
import ltdjms.discord.shared.di.JDAProvider;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.User;

/** Sends admin DM notifications for orders that require manual dispatch. */
public class ShopAdminNotificationService {

  private static final Logger LOG = LoggerFactory.getLogger(ShopAdminNotificationService.class);

  public void notifyAdminsOrderCreated(
      long guildId, long buyerUserId, Product product, String orderType, String orderReference) {
    if (product == null) {
      return;
    }

    Guild guild = JDAProvider.getJda().getGuildById(guildId);
    if (guild == null) {
      LOG.warn("Guild not found when notifying admins: guildId={}", guildId);
      return;
    }

    String message =
        buildAdminOrderNotification(guild, buyerUserId, product, orderType, orderReference);
    Set<Long> notified = new HashSet<>();

    List<Member> members = guild.getMembers();
    if (members != null) {
      for (Member member : members) {
        if (!isAdmin(member, guild)) {
          continue;
        }
        User adminUser = member.getUser();
        if (adminUser == null || !notified.add(adminUser.getIdLong())) {
          continue;
        }
        sendAdminNotification(adminUser, message);
      }
    }

    long ownerId = 0L;
    try {
      ownerId = guild.getOwnerIdLong();
    } catch (Exception e) {
      LOG.debug(
          "Unable to resolve guild owner id for order notification: guildId={}",
          guild.getIdLong(),
          e);
    }
    final long finalOwnerId = ownerId;
    if (finalOwnerId > 0 && !notified.contains(finalOwnerId)) {
      guild
          .retrieveMemberById(finalOwnerId)
          .queue(
              ownerMember -> {
                User owner = ownerMember.getUser();
                if (owner != null) {
                  sendAdminNotification(owner, message);
                }
              },
              failure ->
                  LOG.debug(
                      "Failed to retrieve guild owner for order notification: guildId={},"
                          + " ownerId={}",
                      guild.getIdLong(),
                      finalOwnerId,
                      failure));
    }
  }

  private void sendAdminNotification(User adminUser, String message) {
    adminUser
        .openPrivateChannel()
        .queue(
            channel -> channel.sendMessage(message).queue(),
            failure ->
                LOG.debug(
                    "Failed to open admin DM for order notification: adminUserId={}",
                    adminUser.getIdLong(),
                    failure));
  }

  private String buildAdminOrderNotification(
      Guild guild, long buyerUserId, Product product, String orderType, String orderReference) {
    StringBuilder builder = new StringBuilder();
    builder.append("📩 有新訂單發起，請儘速派單\n\n");
    builder
        .append("**伺服器：** ")
        .append(guild.getName())
        .append(" (`")
        .append(guild.getId())
        .append("`)\n");
    builder.append("**買家：** <@").append(buyerUserId).append(">\n");
    builder.append("**商品：** ").append(product.name()).append("\n");
    builder.append("**訂單類型：** ").append(orderType).append("\n");
    if (orderReference != null && !orderReference.isBlank()) {
      builder.append("**訂單編號：** `").append(orderReference).append("`\n");
    }
    builder.append("\n請使用 `/dispatch-panel` 進行派單分配。");
    return builder.toString();
  }

  private boolean isAdmin(Member member, Guild guild) {
    if (member == null || guild == null) {
      return false;
    }
    if (member.hasPermission(Permission.ADMINISTRATOR)) {
      return true;
    }
    try {
      return guild.getOwnerIdLong() == member.getIdLong();
    } catch (Exception ignored) {
      return false;
    }
  }
}

File: packages/shop/src/services/shop-admin-notification.service.ts
import type { DiscordRuntimeGateway } from '@ltdjms/shared';
import type { DispatchOrderSnapshot } from '../domain/escort-dispatch-handoff-service.js';
import type { Product } from '../domain/product-types.js';
import pino from 'pino';

/** Minimal discord.js User shape used by this service. */
interface DiscordJsUser {
  id: string;
  send(message: string): Promise<unknown>;
}

/** Minimal discord.js GuildMember shape used by this service. */
interface DiscordJsGuildMember {
  user: DiscordJsUser;
}

/** Minimal discord.js Role shape used by this service. */
interface DiscordJsRole {
  permissions: {
    has(permission: string): boolean;
  };
  members: Map<string, DiscordJsGuildMember>;
}

/** Minimal discord.js Guild shape used by this service. */
interface DiscordJsGuild {
  name: string;
  ownerId: string;
  roles: {
    cache: Map<string, DiscordJsRole> & {
      find(predicate: (role: DiscordJsRole) => boolean): DiscordJsRole | undefined;
    };
  };
  members: {
    fetch(userId: string): Promise<DiscordJsGuildMember>;
  };
}

/** Minimal discord.js Client shape used by this service. */
interface DiscordJsClient {
  guilds: {
    cache: Map<string, DiscordJsGuild>;
  };
  users: {
    fetch(userId: string): Promise<DiscordJsUser>;
  };
}

export class ShopAdminNotificationService {
  private readonly log: pino.Logger;

  constructor(
    private readonly discordRuntimeGateway: DiscordRuntimeGateway,
    logger?: pino.Logger,
  ) {
    this.log = logger ?? pino({ level: 'warn' });
  }

  /**
   * Adapter for the AdminOrderNotifier interface used by FiatOrderPostPaymentWorker (P1-10).
   * Delegates to the existing escort notification builder.
   */
  notifyAdminsOrderCreated(
    guildId: number,
    buyerUserId: number,
    dispatchOrder: DispatchOrderSnapshot,
  ): void {
    if (!dispatchOrder) return;
    const guildName = this.getGuildName(guildId);
    const message = this.buildAdminEscortNotification(
      guildId,
      buyerUserId,
      dispatchOrder,
      guildName ?? undefined,
    );
    this.notifyGuildAdmins(guildId, message);
  }

  /**
   * Notifies admins about a new product order with product details.
   * This method is kept for direct use (e.g., manual dispatch).
   * The FiatOrderPostPaymentWorker uses the 3-parameter overload above.
   */
  notifyAdminsProductOrderCreated(
    guildId: number,
    buyerUserId: number,
    product: Product,
    orderType: string,
    orderReference: string,
  ): void {
    if (!product) return;

    const guildName = this.getGuildName(guildId);
    const message = this.buildAdminOrderNotification(
      guildId,
      buyerUserId,
      product,
      orderType,
      orderReference,
      guildName ?? undefined,
    );
    this.notifyGuildAdmins(guildId, message);
  }

  notifyAdminsEscortOrderCreated(
    guildId: number,
    buyerUserId: number,
    dispatchOrder: DispatchOrderSnapshot,
  ): void {
    if (!dispatchOrder) return;

    const guildName = this.getGuildName(guildId);
    const message = this.buildAdminEscortNotification(
      guildId,
      buyerUserId,
      dispatchOrder,
      guildName ?? undefined,
    );
    this.notifyGuildAdmins(guildId, message);
  }

  private notifyGuildAdmins(guildId: number, message: string): void {
    const client = this.discordRuntimeGateway.requireReadyClient() as DiscordJsClient;
    const selfUserId = this.discordRuntimeGateway.selfUserId();
    const notified = new Set<string>();

    try {
      const guild = client.guilds.cache.get(guildId.toString());
      if (!guild) {
        this.log.warn({ guildId }, 'Guild not found when notifying admins');
        return;
      }

      // Find members with ADMINISTRATOR permission via roles rather than iterating all guild members
      const adminMembers = new Set<DiscordJsGuildMember>();
      const adminRole = guild.roles.cache.find((role) => role.permissions.has('Administrator'));
      if (adminRole) {
        for (const [, member] of adminRole.members) {
          adminMembers.add(member);
        }
      }

      for (const member of adminMembers) {
        const adminUserId = member.user.id;
        if (selfUserId && adminUserId === selfUserId) {
          this.log.debug({ guildId, selfUserId }, 'Skipping bot self when notifying admins');
          continue;
        }
        if (notified.has(adminUserId)) continue;
        notified.add(adminUserId);
        this.sendAdminNotification(member.user, message);
      }

      // Also notify guild owner if not already notified
      const ownerId = guild.ownerId;
      if (ownerId && (!selfUserId || ownerId !== selfUserId) && !notified.has(ownerId)) {
        guild.members.fetch(ownerId).then(
          (ownerMember) => {
            if (ownerMember && ownerMember.user) {
              const ownerUserId = ownerMember.user.id;
              if (selfUserId && ownerUserId === selfUserId) return;
              this.sendAdminNotification(ownerMember.user, message);
            }
          },
          (err: unknown) =>
            this.log.debug(
              { guildId, ownerId, error: err },
              'Failed to retrieve guild owner for order notification',
            ),
        );
      }
    } catch (e) {
      this.log.warn({ guildId, error: e }, 'Failed to notify guild admins');
    }
  }

  private sendAdminNotification(user: DiscordJsUser, message: string): void {
    try {
      user.send(message).catch((err: unknown) => {
        this.log.warn(
          { adminUserId: user.id, error: err },
          'Failed to send admin DM for order notification',
        );
      });
    } catch (e) {
      this.log.warn(
        { adminUserId: user.id, error: e },
        'Failed to open admin DM for order notification',
      );
    }
  }

  private getGuildName(guildId: number): string | null {
    try {
      const client = this.discordRuntimeGateway.requireReadyClient() as DiscordJsClient;
      const guild = client.guilds.cache.get(guildId.toString());
      return guild?.name ?? null;
    } catch {
      return null;
    }
  }

  private buildAdminOrderNotification(
    guildId: number,
    buyerUserId: number,
    product: Product,
    orderType: string,
    orderReference: string,
    guildName?: string,
  ): string {
    const lines: string[] = [];
    lines.push('📩 有新訂單發起，請儘速派單\n');
    if (guildName) {
      lines.push(`**伺服器：** ${guildName}（\`${guildId}\`）`);
    } else {
      lines.push(`**伺服器：** \`${guildId}\``);
    }
    lines.push(`**買家：** <@${buyerUserId}>`);
    lines.push(`**商品：** ${product.name}`);
    lines.push(`**訂單類型：** ${orderType}`);
    if (orderReference) {
      lines.push(`**訂單編號：** \`${orderReference}\``);
    }
    lines.push('\n請使用 `/dispatch-panel` 進行派單分配。');
    return lines.join('\n');
  }

  private buildAdminEscortNotification(
    guildId: number,
    buyerUserId: number,
    order: DispatchOrderSnapshot,
    guildName?: string,
  ): string {
    const lines: string[] = [];
    lines.push('📩 有新護航工作交接，請儘速處理\n');
    if (guildName) {
      lines.push(`**伺服器：** ${guildName}（\`${guildId}\`）`);
    } else {
      lines.push(`**伺服器：** \`${guildId}\``);
    }
    lines.push(`**買家：** <@${buyerUserId}>`);
    lines.push(`**來源類型：** ${this.describeSourceType(order.sourceType ?? null)}`);
    if (order.sourceReference) {
      lines.push(`**來源參考：** \`${order.sourceReference}\``);
    }
    if (order.sourceProductName) {
      lines.push(`**來源商品：** ${order.sourceProductName}`);
    }
    if (order.sourceEscortOptionCode) {
      lines.push(`**護航選項：** \`${order.sourceEscortOptionCode}\``);
    }
    lines.push(`**Dispatch 編號：** \`${order.orderNumber}\``);
    lines.push('\n請使用 `/dispatch-panel` 檢視或後續處理此工作項。');
    return lines.join('\n');
  }

  private describeSourceType(sourceType: string | null): string {
    if (!sourceType) return '未知';
    switch (sourceType) {
      case 'MANUAL':
        return '手動派單';
      case 'CURRENCY_PURCHASE':
        return '貨幣購買';
      case 'FIAT_PAYMENT':
        return '法幣付款';
      default:
        return '未知';
    }
  }
}
